runtime/freelist.c

```diff
@@ -534,60 +348,1384 @@ header_t *caml_fl_merge_block (value bp)
+    }
+  }else{
+    b = n->next;
+    n->next = b->next;
+    b->next->prev = n;
+    result = bf_split (wosz, (value) b);
+    /* TODO: splay at [n] if the remainder is smaller than [wosz] */
+    bf_insert_fragment ((value) b);
+    return result;
+  }
+}
+
+static header_t *bf_allocate_from_tree (mlsize_t wosz)
+{
+  large_free_block **n;
+  mlsize_t bound;
+
+  n = bf_search_best (wosz, &bound);
+  if (n == NULL) return NULL;
+  return bf_alloc_from_large (wosz, n, bound);
+}
+
+static header_t *bf_allocate (mlsize_t wosz)
+{
+  value block;
+  header_t *result;
+
+  CAMLassert (sizeof (char *) == sizeof (value));
+  CAMLassert (wosz >= 1);
+
+#ifdef CAML_INSTR
+  if (wosz < 10){
+    ++instr_size[wosz];
+  }else if (wosz < 100){
+    ++instr_size[wosz/10 + 9];
+  }else{
+    ++instr_size[19];
+  }
+#endif /* CAML_INSTR */
+
+  if (wosz <= BF_NUM_SMALL){
+    if (bf_small_fl[wosz].free != Val_NULL){
+      /* fast path: allocate from the corresponding free list */
+      block = bf_small_fl[wosz].free;
+      if (bf_small_fl[wosz].merge == &Next_small (bf_small_fl[wosz].free)){
+        bf_small_fl[wosz].merge = &bf_small_fl[wosz].free;
+      }
+      bf_small_fl[wosz].free = Next_small (bf_small_fl[wosz].free);
+      if (bf_small_fl[wosz].free == Val_NULL) unset_map (wosz);
+      caml_fl_cur_wsz -= Whsize_wosize (wosz);
+#if FREELIST_DEBUG
+      bf_check ();
+#endif
+      return Hp_val (block);
+    }else{
+      /* allocate from the next available size */
+      mlsize_t s = ffs (bf_small_map & ~ ((1 << wosz) - 1));
+#if FREELIST_DEBUG
+      bf_check ();
+#endif
+      if (s != 0){
+        block = bf_small_fl[s].free;
+        CAMLassert (block != Val_NULL);
+        if (bf_small_fl[s].merge == &Next_small (bf_small_fl[s].free)){
+          bf_small_fl[s].merge = &bf_small_fl[s].free;
+        }
+        bf_small_fl[s].free = Next_small (bf_small_fl[s].free);
+        if (bf_small_fl[s].free == Val_NULL) unset_map (s);
+        result = bf_split_small (wosz, block);
+        if (s > Whsize_wosize (wosz)) bf_insert_fragment_small (block);
+#if FREELIST_DEBUG
+        bf_check ();
+#endif
+        return result;
+      }
+    }
+    /* Failed to find a suitable small block: allocate from the tree. */
+    {
+      header_t *big, *p, *q;
+      int i;
+      /* First, try allocating a run of blocks to preload the freelist. */
+      big = bf_allocate_from_tree
+              (Wosize_whsize (BF_PREALLOC_NUM * Whsize_wosize (wosz)));
+      if (big != NULL){
+        p = big;
+        /* Chain the blocks. */
+        i = 0;
+        while (1){
+          *p = Make_header (wosz, 0, Caml_blue);
+          q = (header_t *) &Field (Val_hp (p), wosz);
+          Next_small (Val_hp (p)) = Val_hp (q);
+          ++i;
+          if (i >= BF_PREALLOC_NUM) break;
+          p = q;
+        }
+        Next_small (Val_hp (p)) = Val_NULL;
+        /* Put the chain in the free list. */
+        bf_small_fl[wosz].free = Next_small (Val_hp (big));
```

Comment:
This assumes that `bf_small_fl[wosz]` is currently empty. However, the call to `bf_allocate_from_tree` above on line 1533 could have added a node to it. This leaks a node and results in an assertion failure with the debug runtime.

Reply:
Good catch indeed!

Do you have a simple repro case? I have a fix but I'd like to test it.

Reply:
Afraid not. I think it is causing a segfault in [https://github.com/OCamlPro/ocaml-benchs/blob/master/chameneos-redux/chameneos_redux_evtchn.ml](this benchmark) about 1 time in 20. I'm just checking at the moment that this is indeed the cause.